src/main/java/org/jenkinsci/plugins/workflow/graphanalysis/ChunkFinder.java

```diff
@@ -6,13 +6,34 @@
 import javax.annotation.Nonnull;
 
 /**
- * Used to define the start and end of a {@link FlowChunk} to split a {@link org.jenkinsci.plugins.workflow.flow.FlowExecution}
- * (For use with a {@link SimpleChunkVisitor} in the {@link ForkScanner#visitSimpleChunks(SimpleChunkVisitor, ChunkFinder)}
+ * Think of this as setting conditions to mark a region of interest in the graph of {@link FlowNode} from a {@link org.jenkinsci.plugins.workflow.flow.FlowExecution}.
+ * <p>This is used to define a linear "chunk" from the graph of FlowNodes returned by a {@link ForkScanner}, after it applies ordering.
+ * <p>This is done by invoking {@link ForkScanner#visitSimpleChunks(SimpleChunkVisitor, ChunkFinder)}.
+ * <p>Your {@link SimpleChunkVisitor} will receive callbacks about chunk boundaries on the basis of the ChunkFinder.
+ *   It isresponsible for tracking the state based on events fired
+ *
+ * <p><p><em>Common uses:</em>
+ * <ul>
+ *     <li>Find all nodes within a specific block type, such the block created by a timeout block, 'node' (executor) block, etc</li>
+ *     <li>Find all nodes between specific markers, such as labels, milestones, or steps generating an error</li>
```

Comment:
It's necessary for legacy stages, these are just suggestions about other ways to take advantage of that.  I think milestones might actually be useful -- and in this case we needn't guarantee tree completeness.  

This is for marking a sequence of nodes that comprise an interesting chunk -- handling of the tree itself and any further parsing is delegated to the `SimpleChunkVisitor` implementation.  This decouples concern.